src/Cardano/AddressDerivation.hs

```diff
@@ -202,3 +223,48 @@ class GenMasterKey (key :: Depth -> * -> *) where
 
     -- | Generate a root key from a corresponding seed.
     genMasterKey :: GenMasterKeyFrom key -> ScrubbedBytes -> key 'RootK XPrv
+
+
+-- | Address is mere wrapper around a ByteString and represents an encoded address.
+newtype Address = Address
+    { unAddress :: ByteString
+    } deriving (Show, Generic, Eq, Ord)
+
+instance NFData Address
+
+-- | Available network options.
+data NetworkDiscriminant = Mainnet | Testnet Nat
+
+class NetworkDiscriminantVal (n :: NetworkDiscriminant) where
+    networkDiscriminantVal :: Text
+
+instance NetworkDiscriminantVal 'Mainnet where
+    networkDiscriminantVal =
+        "mainnet"
+
+instance KnownNat pm => NetworkDiscriminantVal ('Testnet pm) where
+    networkDiscriminantVal =
+        "testnet (" <> T.pack (show $ natVal $ Proxy @pm) <> ")"
+
+-- | Magic constant associated to a given network
+newtype ProtocolMagic = ProtocolMagic { getProtocolMagic :: Int32 }
+    deriving (Generic, Show, Eq)
+
+-- | Hard-coded protocol magic for the Byron MainNet
+mainnetMagic :: ProtocolMagic
+mainnetMagic =  ProtocolMagic 764824073
+
+-- | Derive testnet magic from a type-level Nat
+testnetMagic :: forall pm. KnownNat pm => ProtocolMagic
+testnetMagic = ProtocolMagic $ fromIntegral $ natVal $ Proxy @pm
+
+-- | Encoding of addresses for certain key types and backend targets.
+class PaymentAddress (network :: NetworkDiscriminant) key where
```

Comment:
I think we could better organize this instead of putting everything in `Cardano.AddressDerivation`. What about having:


```
src/
└── Cardano
    ├── Address
    │   ├── Derivation.hs
    │   ├── Derivation
    │   │    ├── Byron.hs
    │   │    └── Icarus.hs
    │   └── Payment.hs
    └── Mnemonic.hs
```

Reply:
And putting bits relevant to the construction of Payment address in `Cardano.Address.Payment`